tracing-opentelemetry-jdbc/src/main/java/io/micronaut/tracing/opentelemetry/instrument/jdbc/JdbcTelemetryFactory.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.tracing.opentelemetry.instrument.jdbc;
+
+import io.micronaut.context.annotation.Factory;
+import io.micronaut.context.event.BeanCreatedEventListener;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.Order;
+import io.opentelemetry.api.OpenTelemetry;
+import io.micronaut.jdbc.DataSourceResolver;
+import io.opentelemetry.instrumentation.jdbc.datasource.JdbcTelemetry;
+import jakarta.inject.Singleton;
+import io.micronaut.core.order.Ordered;
+
+import javax.sql.DataSource;
+
+/**
+ * Jdbc telemetry factory.
+ */
+@Factory
+@Internal
+final class JdbcTelemetryFactory {
+
+    /**
+     * Wraps the DataSource so OTEL can gather data for spans.
+     *
+     * @param telemetry instance of {@link OpenTelemetry}
+     * @param resolver the {@link DataSourceResolver}
+     * @return even listener for @{@link DataSource} creation
+     */
+    @Singleton
+    @Order(value = Ordered.HIGHEST_PRECEDENCE)
+    BeanCreatedEventListener<DataSource> otel(
+        OpenTelemetry telemetry,
+        DataSourceResolver resolver,
+        JdbcTelemetryConfiguration jdbcTelemetryConfiguration) {
+        return event -> {
+            DataSource dataSource = event.getBean();
+            return JdbcTelemetry.builder(telemetry)
+                .setDataSourceInstrumenterEnabled(jdbcTelemetryConfiguration.dataSourceInstrumenterEnabled() == null ? Boolean.TRUE : jdbcTelemetryConfiguration.dataSourceInstrumenterEnabled())
+                .setStatementInstrumenterEnabled(jdbcTelemetryConfiguration.statementInstrumenterEnabled() == null ? Boolean.TRUE : jdbcTelemetryConfiguration.statementInstrumenterEnabled())
+                .setStatementSanitizationEnabled(jdbcTelemetryConfiguration.statementSanitizationEnabled() == null ? Boolean.TRUE : jdbcTelemetryConfiguration.statementSanitizationEnabled())
+                .build().wrap(resolver.resolve(dataSource));
```

Comment:
remove the unwrapping since it shouldn't be necessary if the data source is not wrapped

Reply:
removing the resolver the test start falling with

```
Message: No current connection present. Consider declaring @Connectable or @Transactional on the surrounding method
Path Taken: new SchemaGenerator(List<DataJdbcConfiguration E> configurations,JdbcSchemaHandler schemaHandler)
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [io.micronaut.data.jdbc.config.SchemaGenerator] could not be loaded: Error instantiating bean of type  [io.micronaut.data.jdbc.config.SchemaGenerator]
```

Reply:
I think resolver (default one is `DelegatingDataSourceResolver`) should figure out if unwrapping is needed.

Reply:
there shouldn't be any unwrapping. If you unwrap then the datasources bean is now invalid and `@Connectable` will break. The event listener has to run in a phase with the unwrapped datasource